### PR TITLE
iDeep: F.lstm and F.tree_lstm

### PR DESCRIPTION
**Pull request by niboshi**
iDeep support in F.lstm and F.tree_lstm (Separated from #3865. Based on #4276.)

### CODE REVIEW

**Comment by niboshi** [comment]
Rebased.

File: chainer/functions/activation/lstm.py [hunk 1]
```diff
@@ -82,15 +84,19 @@ def forward(self, inputs):
...
         batch = len(x)
 
-        if isinstance(x, numpy.ndarray):
-            a = numpy.tanh(a)
-            i = _sigmoid(i)
-            f = _sigmoid(f)
-            o = _sigmoid(o)
+        if isinstance(x, chainer.get_cpu_array_types()):
+            if intel64.should_use_ideep('>=auto'):
+                xp = intel64.ideep.get_array_module(x)
```
Comment (okuta) [inline chainer/functions/activation/lstm.py]: Why don't you use `xp = intel64.ideep` ?

File: chainer/functions/activation/lstm.py [hunk 2]
```diff
@@ -136,11 +142,15 @@ def forward(self, inputs):
...
         if xp is numpy:
-            tanh_a = xp.tanh(a)
-            sig_i = _sigmoid(i)
-            sig_f = _sigmoid(f)
-            sig_o = _sigmoid(o)
-
+            if intel64.should_use_ideep('>=auto'):
+                xp = intel64.ideep.get_array_module(x)
+            # TODO(iDeep) replace numpy with xp
+            tanh_a = numpy.tanh(a)
```
Comment (okuta) [inline chainer/functions/activation/lstm.py]: Please use `xp.tanh`.

Comment (niboshi) [reply chainer/functions/activation/lstm.py]: Ah, it seems old TODO when ideep did not support it.
Fixed.

**Pull request merged**